File: src/http/h2_client/Stream.zig
//! One in-flight request on a multiplexed HTTP/2 `ClientSession`. Owned by the
//! session's `streams` map; `client` is a weak back-pointer to the `HTTPClient`
//! that the request belongs to (cleared before any terminal callback so the
//! deliver loop never dereferences a freed client).

pub const new = bun.TrivialNew(@This());

id: u31,
session: *ClientSession,
client: ?*HTTPClient,

/// HEADERS + CONTINUATION fragments, decoded once END_HEADERS arrives.
header_block: std.ArrayListUnmanaged(u8) = .{},
/// DATA payload accumulated across one onData() pass.
body_buffer: std.ArrayListUnmanaged(u8) = .{},

/// HPACK is decoded eagerly at parse time so the dynamic table stays
/// consistent across multiple HEADERS in one read; the resulting strings
/// land here until `deliverStream` hands them to handleResponseMetadata.
decoded_bytes: std.ArrayListUnmanaged(u8) = .{},
decoded_headers: std.ArrayListUnmanaged(picohttp.Header) = .{},
/// Final (non-1xx) status code; 0 until the response HEADERS arrive.
status_code: u32 = 0,

state: State = .open,
/// `.closed` was reached via RST_STREAM (sent or received). Kept distinct
/// from `state` so `rst()` stays idempotent (never answers an inbound RST,
/// per §5.4.2) and so RST(NO_ERROR) can be told apart from a clean close.
rst_done: bool = false,
/// Set once a non-1xx HEADERS block has been decoded and is awaiting
/// delivery. Subsequent HEADERS are trailers and decoded-then-dropped.
headers_ready: bool = false,
headers_end_stream: bool = false,
/// Expect: 100-continue is in effect: hold the request body until a 1xx
/// or final status arrives.
awaiting_continue: bool = false,
fatal_error: ?anyerror = null,
/// DATA bytes received since the last per-stream WINDOW_UPDATE. For
/// consumers without `body_consumption_tracked` set this alone drives the
/// credit; for tracked consumers it is the ceiling on what
/// `consumed_bytes` may release.
unacked_bytes: u32 = 0,
/// Bytes the JS `ReadableStream` reader has actually drained, reported via
/// `scheduleResponseBodyConsumed`. Only consulted when
/// `body_consumption_tracked` is true; `replenishWindow` credits
/// `min(consumed_bytes, unacked_bytes)` so a stalled reader withholds the
/// per-stream window and a compressed body can't over-credit.
consumed_bytes: u32 = 0,
/// Σ DATA payload bytes (post-padding) for §8.1.1 Content-Length check —
/// `total_body_received` is clamped at content_length so it can't catch
/// overshoot.
data_bytes_received: u64 = 0,
/// Per-stream send window (server's INITIAL_WINDOW_SIZE plus any
/// WINDOW_UPDATEs minus DATA bytes already framed).
send_window: i32,
/// Unsent suffix of a `.bytes` request body, parked while the send
/// window is exhausted. Borrows from `client.state.request_body`.
pending_body: []const u8 = "",

/// RFC 9113 §5.1. A `Stream` is created by sending HEADERS, so it starts
/// `.open`; `idle`/`reserved` are never represented as objects. END_STREAM
/// half-closes one side; both, or any RST_STREAM, transitions to `.closed`.
pub const State = enum(u2) {
    open,
    /// We have written END_STREAM; no more DATA may be queued.
    half_closed_local,
    /// Peer has sent END_STREAM; further DATA is STREAM_CLOSED.
    half_closed_remote,
    closed,
};

pub fn deinit(this: *@This()) void {
    _ = H2.live_streams.fetchSub(1, .monotonic);
    this.header_block.deinit(bun.default_allocator);
    this.body_buffer.deinit(bun.default_allocator);
    this.decoded_bytes.deinit(bun.default_allocator);
    this.decoded_headers.deinit(bun.default_allocator);
    bun.destroy(this);
}

pub fn rst(this: *@This(), code: wire.ErrorCode) void {
    if (this.rst_done or this.state == .closed) return;
    this.rst_done = true;
    this.state = .closed;
    var value: u32 = @byteSwap(@intFromEnum(code));
    this.session.writeFrame(.HTTP_FRAME_RST_STREAM, 0, this.id, std.mem.asBytes(&value));
}

pub fn sentEndStream(this: *@This()) void {
    this.state = switch (this.state) {
        .open => .half_closed_local,
        .half_closed_remote => .closed,
        else => this.state,
    };
}

pub fn recvEndStream(this: *@This()) void {
    this.state = switch (this.state) {
        .open => .half_closed_remote,
        .half_closed_local => .closed,
        else => this.state,
    };
}

/// We have sent END_STREAM (or RST): no more request DATA may be queued.
pub inline fn localClosed(this: *const @This()) bool {
    return this.state == .half_closed_local or this.state == .closed;
}

/// Peer has sent END_STREAM (or RST): the response body is complete and
/// further inbound DATA is a protocol error.
pub inline fn remoteClosed(this: *const @This()) bool {
    return this.state == .half_closed_remote or this.state == .closed;
}

const ClientSession = @import("./ClientSession.zig");
const H2 = @import("../H2Client.zig");
const std = @import("std");
const wire = @import("../H2FrameParser.zig");

const bun = @import("bun");
const HTTPClient = bun.http;
const picohttp = bun.picohttp;

File: src/http/h2_client/Stream.rs
//! One in-flight request on a multiplexed HTTP/2 `ClientSession`. Owned by the
//! session's `streams` map (as an `Rc`, so frame handlers can hold one while
//! the session's other state is in use); `client` is a back-reference to the
//! request the stream belongs to (cleared before any terminal callback so the
//! deliver loop never touches a finished request).

use core::cell::{Cell, RefCell};
use core::sync::atomic::Ordering;

use crate::Error;
use crate::RequestRef;
use bun_picohttp as picohttp;

pub struct Stream {
    // HTTP/2 stream IDs are 31-bit; the top bit must stay clear.
    pub(crate) id: u32,
    /// Snapshot of `client.async_http_id` at attach time. Stored so the
    /// session's `by_http_id` index can be maintained after `client` has been
    /// cleared (terminal delivery nulls `client` before `remove_stream`).
    /// `None` for clients without an abort-signal store (not indexed).
    pub(crate) async_http_id: Option<u32>,
    /// Cleared before terminal callbacks.
    pub(crate) client: Cell<Option<RequestRef>>,

    /// HEADERS + CONTINUATION fragments, decoded once END_HEADERS arrives.
    pub(crate) header_block: RefCell<Vec<u8>>,
    /// DATA payload accumulated across one onData() pass.
    pub(crate) body_buffer: RefCell<Vec<u8>>,

    /// HPACK is decoded eagerly at parse time so the dynamic table stays
    /// consistent across multiple HEADERS in one read; the resulting strings
    /// land here until `deliverStream` hands them to handleResponseMetadata.
    pub(crate) decoded_bytes: RefCell<Vec<u8>>,
    /// Each header points into `decoded_bytes`, which is not touched again
    /// once these are built.
    pub(crate) decoded_headers: RefCell<Vec<picohttp::Header>>,
    /// Final (non-1xx) status code; 0 until the response HEADERS arrive.
    pub(crate) status_code: Cell<u32>,

    pub(crate) state: Cell<State>,
    /// `.closed` was reached via RST_STREAM (sent or received). Kept distinct
    /// from `state` so `rst()` stays idempotent (never answers an inbound RST,
    /// per §5.4.2) and so RST(NO_ERROR) can be told apart from a clean close.
    pub(crate) rst_done: Cell<bool>,
    /// Set once a non-1xx HEADERS block has been decoded and is awaiting
    /// delivery. Subsequent HEADERS are trailers and decoded-then-dropped.
    pub(crate) headers_ready: Cell<bool>,
    pub(crate) headers_end_stream: Cell<bool>,
    /// Expect: 100-continue is in effect: hold the request body until a 1xx
    /// or final status arrives.
    pub(crate) awaiting_continue: Cell<bool>,
    pub(crate) fatal_error: Cell<Option<Error>>,
    /// DATA bytes consumed since the last WINDOW_UPDATE for this stream.
    pub(crate) unacked_bytes: Cell<u32>,
    /// Σ DATA payload bytes (post-padding) for §8.1.1 Content-Length check —
    /// `total_body_received` is clamped at content_length so it can't catch
    /// overshoot.
    pub(crate) data_bytes_received: Cell<u64>,
    /// Per-stream send window (server's INITIAL_WINDOW_SIZE plus any
    /// WINDOW_UPDATEs minus DATA bytes already framed).
    pub(crate) send_window: Cell<i32>,
    /// Unsent suffix of a `.bytes` request body, parked while the send
    /// window is exhausted. Points into `client.state.request_body`'s
    /// backing, which the request keeps alive while the stream is attached.
    pub(crate) pending_body: Cell<bun_ptr::RawSlice<u8>>,
}

/// RFC 9113 §5.1. A `Stream` is created by sending HEADERS, so it starts
/// `.open`; `idle`/`reserved` are never represented as objects. END_STREAM
/// half-closes one side; both, or any RST_STREAM, transitions to `.closed`.
#[repr(u8)]
#[derive(Copy, Clone, PartialEq, Eq)]
pub enum State {
    Open,
    /// We have written END_STREAM; no more DATA may be queued.
    HalfClosedLocal,
    /// Peer has sent END_STREAM; further DATA is STREAM_CLOSED.
    HalfClosedRemote,
    Closed,
}

impl Drop for Stream {
    fn drop(&mut self) {
        let _ = super::LIVE_STREAMS.fetch_sub(1, Ordering::Relaxed);
    }
}

impl Stream {
    pub(crate) fn new(
        id: u32,
        async_http_id: Option<u32>,
        client: Option<RequestRef>,
        send_window: i32,
    ) -> Self {
        let _ = super::LIVE_STREAMS.fetch_add(1, Ordering::Relaxed);
        Self {
            id,
            async_http_id,
            client: Cell::new(client),
            header_block: RefCell::new(Vec::new()),
            body_buffer: RefCell::new(Vec::new()),
            decoded_bytes: RefCell::new(Vec::new()),
            decoded_headers: RefCell::new(Vec::new()),
            status_code: Cell::new(0),
            state: Cell::new(State::Open),
            rst_done: Cell::new(false),
            headers_ready: Cell::new(false),
            headers_end_stream: Cell::new(false),
            awaiting_continue: Cell::new(false),
            fatal_error: Cell::new(None),
            unacked_bytes: Cell::new(0),
            data_bytes_received: Cell::new(0),
            send_window: Cell::new(send_window),
            pending_body: Cell::new(bun_ptr::RawSlice::EMPTY),
        }
    }

    #[inline]
    pub(crate) fn set_fatal_error(&self, err: Error) {
        self.fatal_error.set(Some(err));
    }

    pub(crate) fn sent_end_stream(&self) {
        self.state.set(match self.state.get() {
            State::Open => State::HalfClosedLocal,
            State::HalfClosedRemote => State::Closed,
            other => other,
        });
    }

    pub(crate) fn recv_end_stream(&self) {
        self.state.set(match self.state.get() {
            State::Open => State::HalfClosedRemote,
            State::HalfClosedLocal => State::Closed,
            other => other,
        });
    }

    /// We have sent END_STREAM (or RST): no more request DATA may be queued.
    #[inline]
    pub(crate) fn local_closed(&self) -> bool {
        let state = self.state.get();
        state == State::HalfClosedLocal || state == State::Closed
    }

    /// Peer has sent END_STREAM (or RST): the response body is complete and
    /// further inbound DATA is a protocol error.
    #[inline]
    pub(crate) fn remote_closed(&self) -> bool {
        let state = self.state.get();
        state == State::HalfClosedRemote || state == State::Closed
    }
}
